### PR TITLE
Fixes several issues in the reactor related components

### PR DESCRIPTION
**Pull request by sjkwak**
## Description
This pull request contains the following changes.

1) Drain pending tasks when recreating the reactor to make sure pending calls get complete.
2) Fix the timeout issue on the close of the receiver and sender.
3) Make session open timeout configurable and use the value of OperationTimeout.
4) When closing the reactor dispatcher, process work items prior to closing the channel.
5) Improve tracing for reactor related components.
6) refactoring codes

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] **I have read the [contribution guidelines](./CONTRIBUTING.md).**
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR.
- [ ] The pull request does not introduce breaking changes (unless a major version change occurs in the assembly and module).
- [ ] If applicable, the public code is properly documented.
- [ ] Pull request includes test coverage for the included changes.
- [ ] The code builds without any errors.

### CODE REVIEW

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/BaseLinkHandler.java [hunk 1]
```diff
@@ -25,26 +25,38 @@ public BaseLinkHandler(final AmqpLink amqpLink) {
     public void onLinkLocalClose(Event event) {
         final Link link = event.getLink();
         if (TRACE_LOGGER.isInfoEnabled()) {
-            TRACE_LOGGER.info(String.format("linkName[%s]", link.getName()));
+            TRACE_LOGGER.info(String.format("onLinkLocalClose linkName[%s]", link.getName()));
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/BaseLinkHandler.java]: >onLinkLocalClose [](start = 45, length = 16)

we removed these at some point. having caller method names can be configured in slf4j - for instance logback configfile takes in %m - to spew out method name..

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java [hunk 2]
```diff
@@ -707,12 +706,13 @@ public ErrorContext getContext() {
                     public void onEvent() {
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java]: >onEvent [](start = 32, length = 7)

what is the motivation behind this change?
This is wrong.

You are effectively scheduling a receiveLink close call and then cancelling the close-timer - without receiving response from the service.

please remember receiveLink.close() doesn't close the link immediately. It schedules a task on an internal reactor queue - for which amqp bytes will be written. We will get a callback with LinkHandler.onLinkRemoteClose, that is when the link is actually closed. Please see onError method in this file.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageSender.java [hunk 3]
```diff
@@ -893,12 +889,13 @@ public void run() {
                     public void onEvent() {
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageSender.java]: >onEvent [](start = 32, length = 7)

this change is same as messagereceiver change.. marking here.. to be sure...

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java [hunk 4]
```diff
@@ -135,26 +135,26 @@ public void run(Selectable selectable) {
     private final class CloseHandler implements Callback {
         @Override
         public void run(Selectable selectable) {
+            workScheduler.run(null);
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java]: >workScheduler.run(null); [](start = 12, length = 24)

why is the order changed ?
The rationale behind old order is that:
after closing the channel - no more writes are allowed, so, it is safer to drain all the tasks.

If tasks are drained - and then, if the channel is closed - won't new tasks be added in between?

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java [hunk 5]
```diff
@@ -300,32 +294,40 @@ private void onReactorError(Exception cause) {
...
+                        "starting new reactor"));
+
+                this.startReactor(new ReactorHandlerWithConnection());
             } catch (IOException e) {
                 TRACE_LOGGER.error(String.format(Locale.US, "messagingFactory[%s], hostName[%s], error[%s]",
                         this.getClientId(), this.getHostName(),
                         ExceptionUtil.toStackTraceString(e, "Re-starting reactor failed with error")));
 
-                this.onReactorError(cause);
+                throw new RuntimeException(e);
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java]: >throw new RuntimeException(e); [](start = 16, length = 30)

Why are we throwing here?
To understand if this is intended - the sideeffect of this is - this wil not bubble up the error to the User. An internal Thread in ExecutorService will abort with this error.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java [hunk 6]
```diff
@@ -456,8 +452,11 @@ public void onError(Exception error) {
                 }
             }
 
-            if (connection != null && connection.getRemoteState() != EndpointState.CLOSED && connection.getLocalState() != EndpointState.CLOSED)
+            if (connection != null && connection.getRemoteState() != EndpointState.CLOSED && connection.getLocalState() != EndpointState.CLOSED) {
                 connection.close();
+            }
+
+            closeTask.complete(null);
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java]: >closeTask.complete(null); [](start = 12, length = 25)

this is wrong. Connection.close() schedules a close call - and then this line notifies users CompletableFuture - that Close is done. But, the truth is its in progress.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java [hunk 7]
```diff
@@ -543,15 +548,42 @@ public void run() {
...
+                    this.rctr.free();
+
                     closeTask.complete(null);
 
-                    if (closeTimer != null)
+                    if (closeTimer != null) {
                         closeTimer.cancel(false);
+                    }
+                } else {
+                    new ScheduledThreadPoolExecutor(1).schedule(new Runnable() {
```
Comment (SreeramGarlapati) [inline azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java]: >new ScheduledThreadPoolExecutor(1) [](start = 20, length = 34)

what is the need for this?

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/BaseLinkHandler.java (hunk 1 above)
Comment (sjkwak) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/BaseLinkHandler.java]: We've seen many cases where log files that customers share don't have method name and hence making it part of log message will be helpful and reduce the chance of asking a repro with updated log setting.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageSender.java (hunk 3 above)
Comment (sjkwak) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageSender.java]: Same comment as above.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java (hunk 4 above)
Comment (sjkwak) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java]: For the most part, close handler will be invoked as part of recreating the reactor due to an error and scheduling a new task will be put into a new reactor and tasks in the old reactor will be drained fully before close handler is invoked. Please note that workScheduler.run reads the pipe first before draining tasks and closed-channel error will be thrown if the channel was already closed.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java (hunk 2 above)
Comment (SreeramGarlapati) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java]: So, to understand your comment: are you proposing an API change - that PartitionReceiver.close() - to be simply fire-and-forget? To understand more:
* Do you also want to reflect this in the API signature - to remove CompletableFuture as a return value for PartitionReceiver.close() method in that case? as there is nothing asynchronous in this notion?
* do you also plan to remove the closeTimer - as it is cancelled as soon as created.?

Today, without this change, if users want to achieve this, they can always fire an async task - PartitionReceiver.close() and not await on the response - so, what is the motivation of this change?

Comment (sjkwak) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java]: Yes, we will need to consider updating API signature to reflect a change, but it won't be necessarily needed as returning a task doesn't have to be combined with a remote I/O operation.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java (hunk 7 above)
Comment (SreeramGarlapati) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java]: The question is that, original code is:
`reactor.free()`
This PR effectively changes this line to 
`ScheduledThreadPoolExecutor(1).schedule(new …..{ reactor.free()  });`

How does this help?
reactor.free is run regardless -  in both the cases!

Also, current eventhub client sdk - takes in an ExecutorService and doesn't spin off any threads in the SDk layer. Please think about similar strategy for this new ScheduledThreadPool.

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java (hunk 6 above)
Comment (SreeramGarlapati) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java]: We fixed several bugs in the past (#5) to clean up all the local resources, by the time close call returns. This change effectively undo's them.
`MessagingFactory` holds on to the `ReactorThread` - until the response from the service is returned. We free up the localresources - like nio.socket & reactor Thread once the service returns or the closeTask times out. 

Again, can you please explain - what is the value add here by removing this functionality. Why can't we let users, who do not care about this - to fire a close() call and not await on the call?

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java (hunk 2 above)
Comment (SreeramGarlapati) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessageReceiver.java]: Ok, so, just to understand, is this behavior change on the Close API - planned across .net clients as well - to fix the close timeout issues due to excessive prefetch? 
or is it more holistic fixing this on the service side?

File: azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java (hunk 4 above)
Comment (SreeramGarlapati) [reply azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/ReactorDispatcher.java]: "For the most part, close handler will be invoked as part of recreating the reactor due to an error"
This statement is wrong! Close handler is invoked every time reactor is free'd - which means this is invoked - when MessagingFactory.close() call is invoked. 

I still did not get your point on, draining tasks and closed channel.
Let me explain this: 
signaling work queue (in the invoke & invoke with delay methods above) is intended to run on a different thread to that of CloseHandler (which runs on Reactor thread). That is the functionality that ReactorDispatcher offers.
So, in this PR, after reversing the order to workScheduler.run() and channel.close(), we are introducing a chance for a race condition - where new task is inserted into the ReactorDispatcher queue - **after** workScheduler.run() finishes and **before** channel.close() is invoked. 
Whereas, the original code handles this case very well.

So, my main question is - please explain what was the motivation - behind reversing the order; was there a customer reported incident - which somehow lead us to this conclusion?

**Pull request merged**